app/mbm/static/js/autocomplete.js

```diff
@@ -0,0 +1,132 @@
+  });
+
+  // TODO: This doesn't really need to depend on geolocation. We want to suggest
+  // results all over the city regardless of where the user is located, so we could
+  // instead hardcode a bounds object for chicago city limits. As a bonus, we could
+  // then set the autocomplete to use strictbounds and only return results within
+  // them rather than just biasing towards those results. Alternatively, we could
+  // have the backend generate these bounds based on where it has route definitions.
+  //
+  // Bias the autocomplete object to the user's geographical location,
+  // as supplied by the browser's 'navigator.geolocation' object.
+  if (navigator.geolocation) {
+    navigator.geolocation.getCurrentPosition(position => {
+      const geolocation = {
+        lat: position.coords.latitude,
+        lng: position.coords.longitude
+      }
+      app.geolocation.handleGPSPositionUpdate(position)
+      const circle = new google.maps.Circle({
+        center: geolocation,
+        // This is 1500 meters for me. Do we really want to bias the autocomplete to locations within 1500 meters?
+        radius: position.coords.accuracy
+      })
+      autocomplete.setBounds(circle.getBounds())
+    })
+  }
+  return autocomplete
+}
+
+
+// This is a very overcomplicated way to put our own items into the
+// autocomplete comboboxes. There doesn't seem to be any official way to do this
+// other than completely implementing the autocomplete widget yourself, thus our hacks.
+// Right now this is only used for a user to select their location.
+//
+// TODO:
+//  - Add user's saved locations to the list and filter them by user input
+//  - Open the dropdown as soon as input is focused (like they do on https://citymapper.com/webapp)
+//  - Allow user to select our options with the keyboard (rn you can only do it with a mouse click or tap)
+//
+const addCustomOption = (app, optionText, callback) => {
+
+  // The google autocomplete container doesn't have any obvious way to find the
+  // input associated with the list of options, which makes it difficult to write
+  // a handler for a user selecting one of our preset options (which input
+  // element should we update?). So instead, we keep track of which input element
+  // received a focus event most recently and update it when a custom option is selected
+  let markerName
+  const recordFocusEvent = (event) => {
+    markerName = event.target.id
+  }
+
+  app.directionsFormElements.source.input.addEventListener('focus', recordFocusEvent)
+  app.directionsFormElements.target.input.addEventListener('focus', recordFocusEvent)
+
+  const ourOption = $(`
+    <div class="pac-item" tabindex="0">
+      <span class="pac-icon pac-icon-marker"></span>
+      <span class="pac-item-query">${optionText}</span>
+    </div>
+  `).on('mousedown', () => {
+    app.directionsFormElements[markerName].input.value = optionText
+    callback(markerName)
+  })
```

Comment:
**[Nitpick, non-blocking]** As a user I feel like I want the `My position` option to display on focus rather than mousedown 🤔 But I recognize that may be difficult to do with the Google autocomplete library so I'm fine punting it for later, just curious if you agree!

Reply:
100% agree about the behavior, and I actually mention that as a TODO at the top. IDK how to make it happen though :upside_down_face:! I'll make an issue for it as a followup.

This `mousedown` event handler is actually for when you click on the option in the (fake) select menu below the input not for when you focus the input. And bizarrely it doesn't seem to work if I bind to the `click` event.

Reply:
https://github.com/jeancochrane/mellow-bike-map/issues/45